# FINITE FIELD ARITHMETIC 
Paling dasar, sebelum group operation (point addition/point doubling)
Mau bikin LUP decomposition (agak kebayang skrg!!)

In [1]:
_.<Z> = GF(2)[]
F.<z> = GF(2^571, modulus=Z^571+Z^10+Z^5+Z^2+1) #[571,10,5,2,0] #degree 571

In [2]:
F.degree()

571

# SQUARING

In [3]:

bil=0
res=0
# s = z
print("0: 1 --> 1")
for i in range(1,F.degree()):
    bil = z^i
    res = bil^2
    print(i,":", bil,"-->", res)

0: 1 --> 1
1 : z --> z^2
2 : z^2 --> z^4
3 : z^3 --> z^6
4 : z^4 --> z^8
5 : z^5 --> z^10
6 : z^6 --> z^12
7 : z^7 --> z^14
8 : z^8 --> z^16
9 : z^9 --> z^18
10 : z^10 --> z^20
11 : z^11 --> z^22
12 : z^12 --> z^24
13 : z^13 --> z^26
14 : z^14 --> z^28
15 : z^15 --> z^30
16 : z^16 --> z^32
17 : z^17 --> z^34
18 : z^18 --> z^36
19 : z^19 --> z^38
20 : z^20 --> z^40
21 : z^21 --> z^42
22 : z^22 --> z^44
23 : z^23 --> z^46
24 : z^24 --> z^48
25 : z^25 --> z^50
26 : z^26 --> z^52
27 : z^27 --> z^54
28 : z^28 --> z^56
29 : z^29 --> z^58
30 : z^30 --> z^60
31 : z^31 --> z^62
32 : z^32 --> z^64
33 : z^33 --> z^66
34 : z^34 --> z^68
35 : z^35 --> z^70
36 : z^36 --> z^72
37 : z^37 --> z^74
38 : z^38 --> z^76
39 : z^39 --> z^78
40 : z^40 --> z^80
41 : z^41 --> z^82
42 : z^42 --> z^84
43 : z^43 --> z^86
44 : z^44 --> z^88
45 : z^45 --> z^90
46 : z^46 --> z^92
47 : z^47 --> z^94
48 : z^48 --> z^96
49 : z^49 --> z^98
50 : z^50 --> z^100
51 : z^51 --> z^102
52 : z^52 --> z^104
53 : z^53 --> z^106
54

543 : z^543 --> z^525 + z^520 + z^517 + z^515
544 : z^544 --> z^527 + z^522 + z^519 + z^517
545 : z^545 --> z^529 + z^524 + z^521 + z^519
546 : z^546 --> z^531 + z^526 + z^523 + z^521
547 : z^547 --> z^533 + z^528 + z^525 + z^523
548 : z^548 --> z^535 + z^530 + z^527 + z^525
549 : z^549 --> z^537 + z^532 + z^529 + z^527
550 : z^550 --> z^539 + z^534 + z^531 + z^529
551 : z^551 --> z^541 + z^536 + z^533 + z^531
552 : z^552 --> z^543 + z^538 + z^535 + z^533
553 : z^553 --> z^545 + z^540 + z^537 + z^535
554 : z^554 --> z^547 + z^542 + z^539 + z^537
555 : z^555 --> z^549 + z^544 + z^541 + z^539
556 : z^556 --> z^551 + z^546 + z^543 + z^541
557 : z^557 --> z^553 + z^548 + z^545 + z^543
558 : z^558 --> z^555 + z^550 + z^547 + z^545
559 : z^559 --> z^557 + z^552 + z^549 + z^547
560 : z^560 --> z^559 + z^554 + z^551 + z^549
561 : z^561 --> z^561 + z^556 + z^553 + z^551
562 : z^562 --> z^563 + z^558 + z^555 + z^553
563 : z^563 --> z^565 + z^560 + z^557 + z^555
564 : z^564 --> z^567 + z^562 + z^

In [4]:
# #versi saved in array
arr = []
bil=0
res=0
# s = z
print("0: 1 --> 1")
for i in range(1,F.degree()):
    bil = z^i
    res = bil^2
#     print(i,":", bil,"-->", res)
    y = str(res).replace("z^","").split(' + ') #.replace("z","1")
    arr.append(y)
#     print(y)
arr 

0: 1 --> 1


[['2'],
 ['4'],
 ['6'],
 ['8'],
 ['10'],
 ['12'],
 ['14'],
 ['16'],
 ['18'],
 ['20'],
 ['22'],
 ['24'],
 ['26'],
 ['28'],
 ['30'],
 ['32'],
 ['34'],
 ['36'],
 ['38'],
 ['40'],
 ['42'],
 ['44'],
 ['46'],
 ['48'],
 ['50'],
 ['52'],
 ['54'],
 ['56'],
 ['58'],
 ['60'],
 ['62'],
 ['64'],
 ['66'],
 ['68'],
 ['70'],
 ['72'],
 ['74'],
 ['76'],
 ['78'],
 ['80'],
 ['82'],
 ['84'],
 ['86'],
 ['88'],
 ['90'],
 ['92'],
 ['94'],
 ['96'],
 ['98'],
 ['100'],
 ['102'],
 ['104'],
 ['106'],
 ['108'],
 ['110'],
 ['112'],
 ['114'],
 ['116'],
 ['118'],
 ['120'],
 ['122'],
 ['124'],
 ['126'],
 ['128'],
 ['130'],
 ['132'],
 ['134'],
 ['136'],
 ['138'],
 ['140'],
 ['142'],
 ['144'],
 ['146'],
 ['148'],
 ['150'],
 ['152'],
 ['154'],
 ['156'],
 ['158'],
 ['160'],
 ['162'],
 ['164'],
 ['166'],
 ['168'],
 ['170'],
 ['172'],
 ['174'],
 ['176'],
 ['178'],
 ['180'],
 ['182'],
 ['184'],
 ['186'],
 ['188'],
 ['190'],
 ['192'],
 ['194'],
 ['196'],
 ['198'],
 ['200'],
 ['202'],
 ['204'],
 ['206'],
 ['208'],
 ['210'],
 ['

# RESULT

In [5]:
print(1, ",", 1)
for i in range(len(arr)):
    for j in range(len(arr[i])):
#         a.append((arr[i][j]))
        if arr[i][j] == "1":
#             print(i,",",0) #utk extract doang
            print(i+2,",",1)#utk langsung ke matlab (index sudah disesuaikan)
        elif arr[i][j] == "z":
#             print(i, "lel")
#             print(i,",",1)  #z becomes 1, 1 becomes 0, then all naik 1 buat ke matlab? # utk extract doang
            print(i+2,",",2) #utk langsung ke matlab (index sudah disesuaikan)
        else:
            print(i+2,",",int((arr[i][j]))+1)

1 , 1
2 , 3
3 , 5
4 , 7
5 , 9
6 , 11
7 , 13
8 , 15
9 , 17
10 , 19
11 , 21
12 , 23
13 , 25
14 , 27
15 , 29
16 , 31
17 , 33
18 , 35
19 , 37
20 , 39
21 , 41
22 , 43
23 , 45
24 , 47
25 , 49
26 , 51
27 , 53
28 , 55
29 , 57
30 , 59
31 , 61
32 , 63
33 , 65
34 , 67
35 , 69
36 , 71
37 , 73
38 , 75
39 , 77
40 , 79
41 , 81
42 , 83
43 , 85
44 , 87
45 , 89
46 , 91
47 , 93
48 , 95
49 , 97
50 , 99
51 , 101
52 , 103
53 , 105
54 , 107
55 , 109
56 , 111
57 , 113
58 , 115
59 , 117
60 , 119
61 , 121
62 , 123
63 , 125
64 , 127
65 , 129
66 , 131
67 , 133
68 , 135
69 , 137
70 , 139
71 , 141
72 , 143
73 , 145
74 , 147
75 , 149
76 , 151
77 , 153
78 , 155
79 , 157
80 , 159
81 , 161
82 , 163
83 , 165
84 , 167
85 , 169
86 , 171
87 , 173
88 , 175
89 , 177
90 , 179
91 , 181
92 , 183
93 , 185
94 , 187
95 , 189
96 , 191
97 , 193
98 , 195
99 , 197
100 , 199
101 , 201
102 , 203
103 , 205
104 , 207
105 , 209
106 , 211
107 , 213
108 , 215
109 , 217
110 , 219
111 , 221
112 , 223
113 , 225
114 , 227
115 , 229
116 , 231
117

444 , 316
445 , 328
445 , 323
445 , 320
445 , 318
446 , 330
446 , 325
446 , 322
446 , 320
447 , 332
447 , 327
447 , 324
447 , 322
448 , 334
448 , 329
448 , 326
448 , 324
449 , 336
449 , 331
449 , 328
449 , 326
450 , 338
450 , 333
450 , 330
450 , 328
451 , 340
451 , 335
451 , 332
451 , 330
452 , 342
452 , 337
452 , 334
452 , 332
453 , 344
453 , 339
453 , 336
453 , 334
454 , 346
454 , 341
454 , 338
454 , 336
455 , 348
455 , 343
455 , 340
455 , 338
456 , 350
456 , 345
456 , 342
456 , 340
457 , 352
457 , 347
457 , 344
457 , 342
458 , 354
458 , 349
458 , 346
458 , 344
459 , 356
459 , 351
459 , 348
459 , 346
460 , 358
460 , 353
460 , 350
460 , 348
461 , 360
461 , 355
461 , 352
461 , 350
462 , 362
462 , 357
462 , 354
462 , 352
463 , 364
463 , 359
463 , 356
463 , 354
464 , 366
464 , 361
464 , 358
464 , 356
465 , 368
465 , 363
465 , 360
465 , 358
466 , 370
466 , 365
466 , 362
466 , 360
467 , 372
467 , 367
467 , 364
467 , 362
468 , 374
468 , 369
468 , 366
468 , 364
469 , 376
469 , 371
469 , 368


# VERIFICATION

In [7]:
(z^570)^2, (z^500)^2

(z^569 + z^18 + z^3 + z^2 + 1, z^439 + z^434 + z^431 + z^429)

contoh cara baca cek di yg gf 2^163 (males nulis lagi di sini)

contoh di gf 2^163
Last element mapping:

(z^162)^2 --> z^161 + z^12 + z^10 + z^5 + z

di index matlab jadi

163 , 162

163 , 13

163 , 11

163 , 6

163 , 2

bener alhamdulillah

## BAWAH INI KOTRETAN

In [17]:
int(arr[2][0])

6

In [18]:

# y.replace(" ","")
y = str(res).replace("z^","").split(' + ') #.replace("z","1")

y

['281', '22', '12', '10', '8', '5', '3']

In [19]:
# t = {}
# group = F.list()
# for i in range(len(group)):
#     if '+' not in str(group[i]):
#         t[group[i]] = (group[i])^2
# #         print(i, group[i],"-->", ((group[i])^2))
# t

In [10]:
for i in range(len(group)):
    if '+' not in str(group[i]):
        t[group[i]] = (group[i])^2
#         print(i, group[i],"-->", ((group[i])^2))
t

NameError: name 'group' is not defined

In [ ]:
# def eval_polynomial(poly, val):
#     xs = [ x.strip().replace('^','**') for x in poly.split('+') ]
#     return sum( [eval(n.replace('x', str(val))) for n in xs] )
listval = list(t.values())
listkey = list(t.keys())

In [ ]:
for i in range(len(t)): #gbs, hrs specify dong (?)
# xs = [ x.strip().replace('^','**') for x in poly.split('+') ]
    xs = [ x.strip().strip('z^') for x in str(listval[i]).split('+') ]
    print(xs)